src/adapter/src/coord.rs

```diff
@@ -6761,6 +6780,127 @@ mod timeline {
             self.advance_to.take()
         }
     }
+
+    /// Interval used to persist durable timestamps. See [`DurableTimestampOracle`] for more
+    /// details.
+    pub static TIMESTAMP_PERSIST_INTERVAL: Lazy<mz_repr::Timestamp> = Lazy::new(|| {
+        Duration::from_secs(15)
+            .as_millis()
+            .try_into()
+            .expect("15 seconds can fit into `Timestamp`")
+    });
+
+    /// A type that wraps a [`TimestampOracle`] and provides durable timestamps. This allows us to
+    /// recover a timestamp that is larger than all previous timestamps on restart. The protocol
+    /// is based on timestamp recovery from Percolator <https://research.google/pubs/pub36726/>. We
+    /// "pre-allocate" a group of timestamps at once, and only durably store the largest of those
+    /// timestamps. All timestamps within that interval can be served directly from memory, without
+    /// going to disk. On restart, we re-initialize the current timestamp to a value one larger
+    /// than the persisted timestamp.
+    ///
+    /// See [`TimestampOracle`] for more details on the properties of the timestamps.
+    pub struct DurableTimestampOracle<T> {
+        timestamp_oracle: TimestampOracle<T>,
+        durable_timestamp: T,
+        persist_interval: T,
+    }
+
+    impl<T: super::CoordTimestamp> DurableTimestampOracle<T> {
+        /// Create a new durable timeline, starting at the indicated time. Timestamps will be
+        /// allocated in groups of size `persist_interval`. Also returns the new timestamp that
+        /// needs to be persisted to disk.
+        ///
+        /// See [`TimestampOracle::new`] for more details.
+        pub async fn new<F, Fut>(
+            initially: T,
+            next: F,
+            persist_interval: T,
+            persist_fn: impl FnOnce(T) -> Fut,
+        ) -> Self
+        where
+            F: Fn() -> T + 'static,
+            Fut: Future<Output = Result<(), crate::catalog::Error>>,
+        {
+            let durable_timestamp = initially.step_forward_by(&persist_interval);
+            persist_fn(durable_timestamp.clone()).await.unwrap();
```

Comment:
It would be really nice to save this closure in the struct and not have to pass it in later methods. Or maybe a closure that looks like `Fn(&mut Catalog, T) -> ()`. However, I cannot figure out how to get this to work properly.